shared/shared.go

```diff
@@ -37,18 +37,20 @@ func ProvingDifficulty(numLabels uint64, k1 uint64) uint64 {
 }
 
 // PowDifficulty returns the target difficulty of finding a nonce in `numLabels` labels.
-// It is calculated such that one computed label is expected to be below the difficulty threshold.
-// The difficulty is calculated as follows:
+// It is calculated such that a high percentage of smeshers find at least one computed label
+// below the difficulty threshold. The difficulty is calculated as follows:
 //
-//	difficulty = 2^256 / numLabels
+//	difficulty = 8 * 2^256 / numLabels
 //
-// TODO(mafa): this difficulty calculation is unfit. There is only a ~ 63% chance to find a nonce
-// in the first `numLabels`	labels and only a ~ 90% chance to find a nonce in the first `2*numLabels` labels.
+// The probability of finding a label below the difficulty threshold within numLabels
+// approaches ~ 99.97% the bigger numLabels gets. Within 1.15 * numLabels the probability
+// approaches 99.99% of finding at least one label below the difficulty threshold.
 func PowDifficulty(numLabels uint64) []byte {
 	difficulty := make([]byte, 33)
 	difficulty[0] = 0x01
 	x := new(big.Int).SetBytes(difficulty)
 	x.Div(x, big.NewInt(int64(numLabels)))
+	x.Lsh(x, 3) // x << 3 == x * 2^3 == x * 8
```

Comment:
The comment wouldn't be necessary if you just multiplied by 8:
```suggestion
	x.Mul(x, 8)
```

Reply:
The reason is speed and memory efficiency:

```
goos: linux
goarch: arm64
pkg: github.com/spacemeshos/post/shared
Benchmark_Lsh-6   	26290420	        45.26 ns/op	     112 B/op	       2 allocs/op
Benchmark_Mul-6   	18627200	        62.79 ns/op	     120 B/op	       3 allocs/op
PASS
ok  	github.com/spacemeshos/post/shared	2.482s
```

Reply:
It's sad that the go stdlib doesn't figure out such optimizations itself. We are back in the '90s (manual clever optimizations that compilers do now).

Btw, is it even a hot path? In other words, does this optimization make a real difference? If not - I'd suggest sticking with the obvious `Mul`.

Reply:
Yes the stdlib seems to not optimize this. I still feel like anyone working on the codebase should know what a bitshift is and that `x <<= 3` is the same as `x *= 8`